pymmmc/transformers.py

```diff
@@ -1,7 +1,58 @@
+import warnings
+
 import aesara.tensor as at
 
 
-def geometric_adstock(x, alpha: float = 0.0, l_max: int = 12, normalize: bool = False):
+def batched_convolution(x, w, axis: int = -1):
+    """Apply a 1D convolution in a vectorized way across multiple batch dimensions.
+
+    Parameters
+    ----------
+    x :
+        The array to convolve.
+    w :
+        The weight of the convolution. The last axis of ``w`` determines the number of steps
+        to use in the convolution.
+    axis : int
+        The axis of ``x`` along witch to apply the convolution
+
+    Returns
+    -------
+    y :
+        The result of convolving ``x`` with ``w`` along the desired axis. The shape of the
+        result will match the shape of ``x`` up to broadcasting with ``w``. The convolved
+        axis will show the results of left padding zeros to ``x`` while applying the
+        convolutions.
+    """
+    orig_ndim = x.ndim
+    axis = axis if axis >= 0 else orig_ndim + axis
+    w = at.as_tensor(w)
+    x = at.moveaxis(x, axis, -1)
+    x_shape, w_shape = at.random.utils.params_broadcast_shapes(
```

Comment:
@ricardoV94, is there another way to implement this shape broadcasting?

- If I’m not explicit about it, broadcasting of symbolic `x` and `w` fails and takes the shape of `x` as the correct one (maybe symbolic isn’t the correct term, I mean tensors that have `None` in their shape).
- When I am explicit but the arrays fail to broadcast at run time, I get segfaults (probably due to the C backend trying to read or write to memory locations that aren’t part of the array)

Reply:
From a closer look at [the source](https://github.com/pymc-devs/pytensor/blob/main/pytensor/tensor/random/utils.py#L55) I think that there should be a couple of assert Ops there.